projects/ngx-json-schema-form/src/lib/layout.service.ts

```diff
@@ -25,147 +26,33 @@ export class LayoutService {
-     *
-     * The returned layout may be longer (or shorter) then the source layout.
-     *
-     * If an item from the source layout returns multiple items (as '*' usually will),
-     * this function will keep all returned items in-line with the surrounding items.
-     *
-     * If an item from the source layout causes an error and returns null, it is
-     * skipped without error, and the function will still return all non-null items.
-     *
-     * @param  layout - the layout to map
-     * @param  function - the funciton to invoke on each element
-     * @param  layoutPointer - the layoutPointer to layout, inside rootLayout
-     * @param  rootLayout - the root layout, which conatins layout
-     * @return the mapped layout
-     */
-    private mapLayout(layout: Array<LayoutItem | string>,
-        fn: (layoutItem: LayoutItem | string, index?: number, pointer?: string, p?: any) => LayoutNode,
-        layoutPointer: string|Array<string> = '', rootLayout: Array<LayoutItem | string> = layout): Array<LayoutNode> {
-        const indexPad = 0;
         const schemaPointers: Map<string, JSONSchema7Definition> = this.schemaService.dataPointerMap;
         const mappedPointers: Set<string> = new Set<string>();
         let starIndex = -1;
 
-        const mappedLayout: Array<LayoutNode> = layout.reduce((currentLayout: Array<LayoutNode>, item: LayoutItem | string, i: number) => {
+        const mappedLayout: Array<LayoutNode> =
+            layout.reduce((currentLayout: Array<LayoutNode>, layoutItem: LayoutItem | string, i: number): Array<LayoutNode> => {
             let newLayout: Array<LayoutNode> = currentLayout;
-            if (<string>item === '*') {
+
+            if (<string>layoutItem === '*') {
                 starIndex = i;
             } else {
-                const realIndex: number = +i + indexPad;
-                const newLayoutPointer = `${layoutPointer}/${realIndex}`;
-                // Note: removed logic to convert tabs to items and items to [items]
-                // if (item.items) {
-                //     newNode.items = LayoutService.mapLayout(item.items, fn, `${newLayoutPointer}/items`, rootLayout);
-                // }
-                const newNode: LayoutNode = fn(clone(item), realIndex, newLayoutPointer, rootLayout);
-                if (newNode) {
-                    if (!newNode.type && isPlainObject(schemaPointers.get(newNode.dataPointer))) {
-                        // TODO: review getInputType() fn and handle missing functionality
-                        newNode.type = <string>(<JSONSchema7>schemaPointers.get(newNode.dataPointer)).type;
-                    }
-                    if (newNode.type) {
-                        mappedPointers.add(newNode.dataPointer);
-                        newLayout = currentLayout.concat(newNode);
-                    }
+                try {
+                    const layoutNode: LayoutNode = LayoutNode.create(layoutItem, this.schemaService);
```

Comment:
Would a service be better here than a static class?  Might be better for testing?

Reply:
The more I think about it, Angular does not seem to provide a clear way to handle this.  I need a factory to create the LayoutNodes, but angular factories work at the DI level to help with creating a dependency.  That doesn't work for me, at I haven't dug deep enough to figure out how. I can export a factory function and inject that, but seems simpler to just stick it on LayoutNode as static and keep it all consolidated.